torch/nn/quantized/modules/linear.py

```diff
@@ -171,3 +171,54 @@ def from_float(mod):
         qlinear.out_scale = torch.tensor([act_qparams[0]])
         qlinear.out_zero_point = torch.tensor([act_qparams[1]])
         return qlinear
+
+
+class DynamicLinear(NNLinear):
+    r"""
+    A dynamic quantized linear module with quantized tensor as inputs
+    and outputs.
+    We adopt the same interface as `torch.nn.Linear`, please see
+    https://pytorch.org/docs/stable/nn.html#torch.nn.Linear for documentation.
+
+    Similar to `torch.nn.Linear`, attributes will be randomly initialized at
+        module creation time and will be overwritten later
+
+    Attributes:
+        weight: the non-learnable quantized weights of the
+                module which are of shape :math:`(\text{out\_features}, \text{in\_features})`.
+        bias:   the non-learnable bias of the module of shape :math:`(\text{out\_features})`.
+                If :attr:`bias` is ``True``, the values are initialized to zero.
+        scale: `scale` parameter of weight Quantized Tensor, type: double
+        zero_point: `zero_point` parameter for weight Quantized Tensor, type: long
+
+    Examples::
+
+        >>> m = nn.quantized.DynamicLinear(20, 30)
+        >>> input = torch.randn(128, 20)
+        >>> output = m(input)
+        >>> print(output.size())
+        torch.Size([128, 30])
+    """
+
+    __constants__ = ['bias', 'in_features', 'out_features', 'scale', 'zero_point']
+
+    def __init__(self, in_features, out_features, bias=True):
+        super(DynamicLinear, self).__init__(in_features, out_features, bias)
+        del self.weight
+        weight_fp32 = torch.Tensor(out_features, in_features).float()
+        self.register_buffer('weight', weight_fp32)
+        del self.bias
+        bias_fp32 = torch.Tensor(out_features).float()
+        self.register_buffer('bias', bias_fp32)
+
+        weight_prepack, col_offsets, self.scale, self.zero_point = torch.fbgemm_linear_quantize_weight(weight_fp32)
+        self.register_buffer(
```

Comment:
this makes this module non-serializable. I think you don't need to store both copies of float and non-float weight and in general it should have the same behavior as regular QLinear (including setstate/getstate). Ideally, if the answer to above questions is "yes", we can even share some common base class

Reply:
We will only keep the non-float weight.